### DISCUSSION TITLE
User Interface

### DISCUSSION THREAD

**Discussion by rafmudaf** [NatLabRockies/floris#464] [v3 Design Discussion]
The user interface to FLORIS has been greatly reduced and in some cases simplified from v2. The majority of the effort in designing v3 was in the core compute code. This discussion outlines considerations for a user interface to v3.

# Turbine grid
The meaning of `turbine_grid_points` in the solver input block is unclear and it is very easy to inadvertently use way too many points. This results in a very slow calculation but with no indication that there are too many points.

Possible solutions:
- Catch this in an input preprocessing / validation function
- Cap the size of the grid
- Change the name of the input field

# Nuances to work out

### Plot horizontal plane without first calculate wake
If you do `fi.reinitialize` and then try to plot the flow field with `fi.calculate_horizontal_plane`, `fi.floris.farm.hub_heights` is 1-dimensional. However, after running `fi.calculate_wake()`, it is 3-dimensional.

```python
fi.reinitialize(
    wind_directions=wind_directions,
    wind_speeds=wind_speeds
)

fi.calculate_wake() # <--- Comment this line to compare

iea15mw_hub_height = fi.floris.farm.hub_heights[0,0,0]
cut_plane = fi.calculate_horizontal_plane(
    height=iea15mw_hub_height,
    wd=[270.0],
    ws=[wind_speeds[2]],
    yaw_angles=-20 * np.ones((1,1,12))
)
```

### Yaw angle shape given to calculate_XX_plane
Since these functions require using only one wind speed and wind direction, the yaw angle array must have shape (1,1,N Turbines). However, in other functions where we calculate wake, we would likely use an array with the full wind direction and wind speed information. This is challenging to understand from the use perspective.

```python

fi.reinitialize(
    wind_directions=wind_directions,
    wind_speeds=wind_speeds
)
fi.calculate_wake()
iea15mw_hub_height = fi.floris.farm.hub_heights[0,0,0]

YAW = -20
aep = fi.get_farm_AEP(
    freq=freq,
    yaw_angles=YAW * np.ones((N_WIND_DIRECTIONS,N_WIND_SPEEDS,N_TURBINES))
) / 1000000
cut_plane = fi.calculate_horizontal_plane(
    height=iea15mw_hub_height,
    wd=[270.0],
    ws=[wind_speeds[2]],
    yaw_angles=YAW * np.ones((1,1,N_TURBINES))  # <-- Replace this with an array like used above
)
wakeviz.visualize_cut_plane(cut_plane)
```

**Comment by paulf81** [comment]
These are all good suggestions, my vote is for:

- short term: add a cap of 3 since I've never used myself or heard yet of someone using more
- medium term: change the name and to clear things up, something like
-     a) Make clear the actual number of points will be the square of this number
-     b) That this is the points per turbine, not total
